source/blog/_posts/2015-08-11-seo-engine.md

```diff
@@ -0,0 +1,242 @@
+sRewriteTable::$replaceRules property
+* `\sRewriteTable::sCreateRewriteTable`: Is the main entry point of the SEO url generation and trigger all of the follwing methods
+* `\sRewriteTable::baseSetup`: Will be called, before the actual SEO URL generation starts. Will initialize the smarty template
+engine with e.g. the Shopware config object, register some additional smarty plugins and also try to set the memory_limit / max_execution_time
+config vars
+* `\sRewriteTable::sCreateRewriteTableCleanup`: This method will remove SEO URLs of products / categories / blogs / static pages etc
+where the referenced entity (e.g. product) is not existing anymore. Shopware will never delete SEO URLs of items, that are still existing.
+* `\sRewriteTable::sCreateRewriteTableStatic`: Will generate the static URLs that can be created in the SEO config. Static
+URLs will allow you to add SEO URLs for custom controllers and landing pages
+* `\sRewriteTable::sCreateRewriteTableCategories`: Will iterate all Shopware categories, apply the category template to them and
+write the resulting SEO URL to the database
+* `\sRewriteTable::sCreateRewriteTableBlog`: Will iterate all Shopware blogs, apply the blog template to them and
+write the resulting SEO URL to the database
+* `\sRewriteTable::sCreateRewriteTableCampaigns`: Will iterate all Shopware shopping worlds, apply the emotion template to them
+and write the resulting SEO URL to the database
+* `\sRewriteTable::sCreateRewriteTableArticles`: Will iterate all Shopware products, apply the product template to them
+and write the resulting SEO URL to the database
+* `\sRewriteTable::sCreateRewriteTableContent`: Will iterate all Shopware forms and static pages, apply the corresponding
+template to them and write the resulting SEO URL to the database
+* `\sRewriteTable::sCreateRewriteTableSuppliers`: Will iterate all Shopware suppliers, apply the corresponding template to them
+and write the resulting SEO URL to the database
+
+## Canonicalize all the links
+After building the SEO links as described above, Shopware will use those SEO links instead of the technical links
+everywhere in the template.
+
+The most common call to build a SEO URL from technical information is:
+
+```
+$query = array(
+    'controller' => 'my-controller',
+    'module' => 'frontend',
+    'action' => 'my-action',
+    'some-param' => 123456,
+);
+
+$url = Shopware()->Router()->assemble($query);
+```
+
+This will internally look up the SEO url for the route `http://my-shop.com/frontend/my-controller/my-action?some-param=123456`
+and print that out, if it is available. When using smarty, you can also use the `url` plugin - it will also take care of proper
+rewriting the URLs:
+
+```
+{url module=frontend controller=my-controller action=my-action some-param=123456}
+```
+
+As it is critical in regards of SEO to make sure, that always SEO routes are shown, Shopware will also find all URLs in the
```

Comment:
As it is critical in terms of SEO to make sure that SEO routes are always shown,